File: Day-31/FindAllDuplicates.py
def FindAllDuplicates(arr):
    map = {}
    returnArray = []

    for i in arr:
        if i in map:
            returnArray.append(i)
        else:
            map[i] = i
    
    return returnArray

print(FindAllDuplicates([4,3,2,7,8,2,3,1]))

def FindAllDuplicatesOneSpace(arr):
    output=[]
    for num in arr:
        if arr[abs(num) - 1] < 0:
            print('if')
            output.append(abs(num))
        else:
            print('else')
            arr[abs(num) - 1] *= -1
    return output

print(FindAllDuplicatesOneSpace([4,3,2,7,8,2,3,1]))

File: Day-31/FindAllDuplicates.js
function AllDuplicatesInArray(arr) {
  let map = {};
  let returnArray = [];
  for (let i = 0; i < arr.length; i++) {
    if (map[arr[i]]) {
      returnArray.push(arr[i]);
    } else {
      map[arr[i]] = true;
    }
  }
  return returnArray;
}

console.log(AllDuplicatesInArray([4, 3, 2, 7, 8, 2, 3, 1]));

function AllDuplicates(arr) {
  let returnArray = [];

  for (let i = 0; i < arr.length; i++) {
    for (let j = 0; j < arr.length; j++) {
      if (i !== j && arr[i] === arr[j] && !returnArray.includes(arr[j])) {
        returnArray.push(arr[j]);
      }
    }
  }
  return returnArray;
}

console.log(AllDuplicates([4, 3, 2, 7, 8, 2, 3, 1]));
